test/common/network/connection_impl_test.cc

```diff
@@ -509,6 +509,86 @@ TEST_P(ConnectionImplTest, ReadDisable) {
   disconnect(false);
 }
 
+// The HTTP/1 codec handles pipelined connections by relying on readDisable(false) resulting in the
+// subsequent request being dispatched. Regression test this behavior.
+TEST_P(ConnectionImplTest, ReadEnableDispatches) {
```

Comment:
Is it possible to have an explicit test for the disconnect case to make sure we don't dispatch on close?

Reply:
We have one!  
If you check out https://github.com/envoyproxy/envoy/pull/8352/commits/a16cdeeb3d273fbfca49ae364a256e958975cbbe
I had to change TEST_P(ConnectionImplTest, EmptyReadOnCloseTest) 
to drain the buffer (or it got an extra dispatch) and added 
TEST_P(ConnectionImplTest, ReadEventIfBufferedDataOnClose) {
to show with that draft (which I didn't like) how we got the event when we had buffered data.

Now that we have a fix which tests for explicit kick, ConnectionImplTest.EmptyReadOnCloseTest shows that when we close when there's buffered data, we don't push it up.

The one exception is if we *explicitly* ask for a kick, and then do a read and get the FIN, we will push the data before we push the close.  I *think* that's the correct behavior, as it means the behavior when we have buffered data and ask for dispatch, matches when there's kernel buffered data, which is to say we always pass up the data before closing.

Reply:
OK, awesome, thanks for clarifying.